### PR TITLE
util: move IntValue to util.optional

### PR DESCRIPTION
**Pull request by RaduBerinde**
This change moves IntValue to optional.Uint and reimplements it as a
proto. The main advantage is that the struct cannot be set to or
compared against a numeric constant, and can't be legally printed
using `%d` (these are very error prone, as they don't compare the
logical value).

Release note: None

### CODE REVIEW

**Review by asubiotto — APPROVED** [review]
:lgtm: I don't love the `optional` package name but I can't come up with something better and naming is hard.

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-:-MMou9Ce9KZggRXhocq2:b4fkwug)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MMotRzZDzB1ctMy2uFa:b-somp7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
> ```Go
> 	resetUint := func(v *optional.Uint) {
> 		if v.HasValue() {
> 			v.Set(0)
> ```

BTW, should we set this to 1 for consistent behavior with time values?
___
*[pkg/util/optional/optional.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMothJV1zFKbFqlW4tN:-MMothJV1zFKbFqlW4tO:b-q12nzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/util/optional/optional.proto#L15)):*
> ```protobuf
> // All the concepts here are "physical plan" concepts.
> 
> syntax = "proto2";
> ```

Can we make this `proto3`? Not sure if we can use a `proto3` message in a `proto2` definition.

**Review by RaduBerinde — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-:-MMscH81EJf8VPiL-VDB:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MMsb3Q3BBmhurTRo6MS:b-rpf7vv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

BTW, should we set this to 1 for consistent behavior with time values?
</blockquote></details>

I don't think it matters much, but we should probably reset times to 0 instead (with your change). Note that in practice, 1ns will show up as 0 anyway. If we reset to a non-zero value, we might want to make it so that 0 stays 0 and any other value gets reset to 1 or whatever.
___
*[pkg/util/optional/optional.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMothJV1zFKbFqlW4tN:-MMsbg1748KX-AT6bbkN:b-90jbmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/util/optional/optional.proto#L15)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can we make this `proto3`? Not sure if we can use a `proto3` message in a `proto2` definition.
</blockquote></details>

Done. It should work both ways.

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-:-MMuODaKBFs8diWHZ4Ft:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MMuO1cM9PcRL30UPJhu:bcy67qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think it matters much, but we should probably reset times to 0 instead (with your change). Note that in practice, 1ns will show up as 0 anyway. If we reset to a non-zero value, we might want to make it so that 0 stays 0 and any other value gets reset to 1 or whatever.
</blockquote></details>

:+1: I think a 0 value should just be elided from the plan as we currently do. Will change the default time in my PR.

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MMuiRrJ-IaEerpWPS4g:briy82m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

:+1: I think a 0 value should just be elided from the plan as we currently do. Will change the default time in my PR.
</blockquote></details>

Then what's the point of an "optional" time value type if you want to treat 0 as unset, like we do already?

**Review by asubiotto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56998)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MMundLGDe-kyCUnA04h:b-p2hja9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Then what's the point of an "optional" time value type if you want to treat 0 as unset, like we do already?
</blockquote></details>

I misspoke above, what I mean by 0 value above is an "unset" value (i.e. `HasValue` returns false) which would allow us to uncover cases where we don't set stats properly. Setting the value to 0 (`HasValue` returns true) to make it deterministic and displaying that on the EXPLAIN ANALYZE plan makes sense to me, so I'll change `time.Duration` to that same behavior in this method. Hope that makes sense :joy:

**Comment by RaduBerinde** [comment]
___
*[pkg/sql/execinfrapb/component_stats.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56998#-MMotRzZDzB1ctMy2uF_:-MN-f3sq5E5rnA_sqj0J:bxzi5zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d748b466de8e901265d2e7536487dc25f990991/pkg/sql/execinfrapb/component_stats.go#L138)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I misspoke above, what I mean by 0 value above is an "unset" value (i.e. `HasValue` returns false) which would allow us to uncover cases where we don't set stats properly. Setting the value to 0 (`HasValue` returns true) to make it deterministic and displaying that on the EXPLAIN ANALYZE plan makes sense to me, so I'll change `time.Duration` to that same behavior in this method. Hope that makes sense :joy: 
</blockquote></details>

Makes sense, we are on the same page!

File: pkg/util/optional/uint.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"fmt"
+	"math"
+)
+
+// MakeUint returns an Uint with a set value. The maximum value that can be
+// represented is (MaxUint64-1).
+func MakeUint(value uint64) Uint {
+	if value == math.MaxUint64 {
+		return Uint{
```
Comment (nvanbenschoten) [inline pkg/util/optional/uint.go:22]: Drive-by comment: should we have made this case panic? Is there a benefit to silently accepting a value that won't round-trip through the optional wrapper?

Comment (RaduBerinde) [reply pkg/util/optional/uint.go:22]: I debated that too. Since currently this is used for stats, I went this way to avoid a potentially unhandled panic.
